SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/SpatialVirtualWorkerTranslator.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "Interop/SpatialReceiver.h"
+#include "SpatialConstants.h"
+#include "Utils/SchemaUtils.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialVirtualWorkerTranslationManager);
+
+SpatialVirtualWorkerTranslationManager::SpatialVirtualWorkerTranslationManager()
+	: bWorkerEntityQueryInFlight(false)
+{}
+
+void SpatialVirtualWorkerTranslationManager::Init(
+	USpatialReceiver* InReceiver,
+	USpatialWorkerConnection* InConnection,
+	SpatialVirtualWorkerTranslator* InTranslator)
+{
+	check(InReceiver != nullptr);
+	Receiver = InReceiver;
+
+	check(InConnection != nullptr);
+	Connection = InConnection;
+
+	check(InTranslator != nullptr);
+	Translator = InTranslator;
+}
+
+void SpatialVirtualWorkerTranslationManager::AddVirtualWorkerIds(const TSet<VirtualWorkerId>& InVirtualWorkerIds)
+{
+	// Currently, this should only be called once on startup. In the future we may allow for more
+	// flexibility. 
+	check(UnassignedVirtualWorkers.IsEmpty());
+	for (VirtualWorkerId VirtualWorkerId : InVirtualWorkerIds)
+	{
+		UnassignedVirtualWorkers.Enqueue(VirtualWorkerId);
+	}
+}
+
+void SpatialVirtualWorkerTranslationManager::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
+{
+	const bool bAuthoritative = AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE;
+
+	if (!bAuthoritative)
+	{
+		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Error, TEXT("Lost authority over the translation mapping. This is not supported."));
+	}
+
+	if (AuthOp.component_id == SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID)
+	{
+		UE_LOG(LogSpatialVirtualWorkerTranslationManager, Log, TEXT("Authority over the VirtualWorkerTranslationManager has changed. This worker %s authority."), bAuthoritative ? TEXT("now has") : TEXT("does not have"));
+
+		if (bAuthoritative)
```

Comment:
If we return inside the `!bAuthoritative` above we can remove this conditional?

Reply:
yup